### PR TITLE
API to fetch observations by ad-hoc flag

### PR DESCRIPTION
**Pull request by tommylau523**

### CODE REVIEW

**Comment by tommylau523** [comment]
* **#2688** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2688?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2688?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2697** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2697?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt [hunk 1]
```diff
@@ -395,12 +395,63 @@ class ObservationsController(
     return SimpleSuccessResponsePayload()
   }
 
+  @GetMapping("/adHoc")
+  @Operation(summary = "Gets a list of observations of planting sites.")
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt]: Including "ad-hoc" in the summary would make it clearer.

File: src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt [hunk 2]
```diff
@@ -395,12 +395,63 @@ class ObservationsController(
...
+    val plotCounts: Map<ObservationId, ObservationPlotCounts>
+
+    if (plantingSiteId != null) {
+      observations =
+          observationStore.fetchObservationsByPlantingSite(plantingSiteId, isAdHoc = true)
+      plantingSites =
+          mapOf(
+              plantingSiteId to
+                  plantingSiteStore.fetchSiteById(plantingSiteId, PlantingSiteDepth.Site))
+      plotCounts = observationStore.countPlots(plantingSiteId)
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt]: Maybe this should take an argument to control whether it's counting plots in ad-hoc or assigned observations so we don't waste database time counting plots for observations we'll ignore.

File: src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt [hunk 3]
```diff
@@ -1059,7 +1114,10 @@ data class ScheduleObservationRequestPayload(
           state = ObservationState.Upcoming)
 }
 
-data class ScheduleAdHocObservationResponsePayload(val id: ObservationId) : SuccessResponsePayload
+data class ScheduleAdHocObservationResponsePayload(
+    val observationId: ObservationId,
+    val plotId: MonitoringPlotId
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt]: May as well toss a trailing comma on here.

**Review by sgrimm — CHANGES_REQUESTED** [review]
Seems good generally, with a couple nits and one change to improve query efficiency.

File: src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt [hunk 4]
```diff
@@ -395,6 +395,57 @@ class ObservationsController(
...
+          description =
+              "Limit results to observations of planting sites in a specific organization. " +
+                  "Ignored if plantingSiteId is specified.")
+      organizationId: OrganizationId? = null,
+      @RequestParam
+      @Schema(
+          description =
+              "Limit results to observations of a specific planting site. Required if " +
+                  "organizationId is not specified.")
+      plantingSiteId: PlantingSiteId? = null,
```
[the remark this answers is not in the document]
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/tracking/api/ObservationsController.kt:412]: Not needed. Springdoc-openapi detects the nullable type automatically.

**Review by sgrimm — APPROVED** [review]

**Comment by tommylau523** [comment]
### Merge activity

* **Dec 17, 4:28 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2688).
* **Dec 17, 4:30 PM EST**: [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2688) couldn't merge this PR because it had conflicts with the trunk branch.
* **Dec 17, 5:05 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/2688).

**Pull request merged**